File: solutions/typescript/023-reverse-nodes-in-k-group.ts
/**
 * Problem 23: Reverse Nodes in k-Group (LeetCode 25)
 * Difficulty: Hard
 * Language: TypeScript
 */
function reverseKGroup(head: ListNode | null, k: number): ListNode | null {
  const toList = (node: ListNode | null): number[] =>
    node === null ? [] : [node.val, ...toList(node.next)];

  const fromList = (arr: number[]): ListNode | null =>
    arr.reduceRight<ListNode | null>((next, val) => new ListNode(val, next), null);

  const items = toList(head);

  const process = (arr: number[]): number[] => {
    if (arr.length < k) return arr;
    const group = arr.slice(0, k).reverse();
    return [...group, ...process(arr.slice(k))];
  };

  return fromList(process(items));
}

File: solutions/rust/023-reverse-nodes-in-k-group.rs
/**
 * Problem 23: Reverse Nodes in k-Group (LeetCode 25)
 * Difficulty: Hard
 * Language: Rust
 */
pub fn reverse_k_group(head: Option<Box<ListNode>>, k: i32) -> Option<Box<ListNode>> {
    let mut vals: Vec<i32> = Vec::new();
    let mut curr = &head;
    while let Some(node) = curr {
        vals.push(node.val);
        curr = &node.next;
    }

    fn process(arr: &[i32], k: usize) -> Vec<i32> {
        if arr.len() < k { return arr.to_vec(); }
        let mut group: Vec<i32> = arr[..k].iter().rev().copied().collect();
        group.extend(process(&arr[k..], k));
        group
    }

    let result = process(&vals, k as usize);
    result.into_iter().rev().fold(None, |next, val| {
        Some(Box::new(ListNode { val, next }))
    })
}
